### PR TITLE
v2.0.remove-unnecessary-inheritance.0

### PR DESCRIPTION
**Pull request by defuse**
See the commit messages.

### CODE REVIEW

**Comment by defuse** [comment]
@paragonie-scott: It would be *wonderful* if you could review this PR, and, at your leisure, argue against removing the inheritance if you think it's valuable.

**Comment by defuse** [comment]
(What it does is written in the commit message)

**Comment by paragonie-scott** [comment]
Aside from the failure caused by #22 (and possibly any errors that failure is hiding), this looks very good to me. I like the way you organize everything. :+1:

**Comment by paragonie-scott** [comment]
>  It would be *wonderful* if you could review this PR, and, at your leisure, argue against removing the inheritance if you think it's valuable.

The only reason I used inheritance in my initial PR #78 was to expose common methods that the derived classes (`Crypto` and `File) could use, but were not public. It's not a feature I feel super strongly about, and someone could always make

    class CryptoProxy extends \Defuse\Crypto\Core {
        public function getInaccessibleMethod($method, $args = []) {
            return parent::$method(...$args);
        }
    }

and there's nothing inheritance can do to stop this.

File: src/File.php [hunk 1]
```diff
@@ -393,28 +416,31 @@ public static function decryptResource($inputHandle, $outputHandle, $key)
         $remaining = 4;
```
Comment (paragonie-scott) [inline src/File.php]: For this, shouldn't we use the header length constant?

File: src/Key.php [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+        $this->config = self::GetKeyVersionConfigFromKeyHeader($this->key_version_header);
+    }
+
+    public function saveToAsciiSafeString()
+    {
+        return Encoding::binToHex(
+            $this->key_version_header .
+            $this->key_bytes .
+            hash(
+                $this->config['CHECKSUM_HASH_FUNCTION'],
```
Comment (paragonie-scott) [inline src/Key.php]: Should this also be converted from an array to an object?

**Comment by paragonie-scott** [comment]
Aside from a few very minor things this looks good to me. I do have one suggestion but I'll open an issue for that.

**Comment by defuse** [comment]
Okay, fixed those minor issues, and the travis testing is passing. I think it's time to land this!

**Pull request merged**